### PR TITLE
Feature/email settings

### PR DESCRIPTION
**Pull request by euglazer**

### CODE REVIEW

**Comment by euglazer** [comment]
description of PR here: https://trello.com/c/9xzheCyk/295-8-email-settings-5

UI partner to this PR here: https://github.com/cobudget/cobudget-ui/pull/264

File: app/services/bucket_service.rb [hunk 1]
```diff
@@ -8,9 +9,11 @@ def self.send_bucket_created_emails(bucket: )
   end
 
   def self.send_bucket_live_emails(bucket: )
-    memberships = bucket.group.memberships.active.reject { |membership| membership.member == bucket.user }
-    memberships.each do |membership|
-      member = membership.member
+    group = bucket.group
+    members = bucket.participants(exclude_author: true, subscribed: true)
+    members.each do |member|
```
Comment (mixmix) [inline app/services/bucket_service.rb]: @euglazer I think you should sort our archived members with a scope, and just iterate over those : 

```ruby 
members.published.each do |member|
```

if you think thats a bad idea, or are unclear how to execute please ask Rob / me

Comment (euglazer) [inline app/services/bucket_service.rb]: That would be ideal, but the archived attribute is bound to the membership, not the member

Comment (euglazer) [inline app/services/bucket_service.rb]: @mixmix

File: db/migrate/20151115221849_add_subscribed_to_participant_activity_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class AddSubscribedToParticipantActivityToUsers < ActiveRecord::Migration
+  def up
+    add_column :users, :subscribed_to_participant_activity, :boolean, default: false
+    User.update_all(subscribed_to_participant_activity: false)
```
Comment (mixmix) [inline db/migrate/20151115221849_add_subscribed_to_participant_activity_to_users.rb]: I think you should add a validation to the `User` model for these fields ... as in just check they're presrent

File: spec/services/bucket_service_spec.rb [hunk 3]
```diff
@@ -5,17 +5,106 @@
...
+      # create an archived participant
+      @archived_participant = create_bucket_participant(bucket: bucket, subscribed: true)
+      Membership.find_by(group: bucket.group, member: @archived_participant).update(archived_at: DateTime.now.utc - 5.days)
+
+      BucketService.send_bucket_live_emails(bucket: bucket)
+      @emails = ActionMailer::Base.deliveries
+      @email_recipients = @emails.map { |e| e.to.first }
+    end
+
+    it "sends emails to all bucket participants subscribed to participant activity, whose membership in the group is active, except bucket author" do
```
Comment (mixmix) [inline spec/services/bucket_service_spec.rb]: this is all really clear. 
nice one

File: spec/services/bucket_service_spec.rb [hunk 4]
```diff
@@ -5,17 +5,106 @@
...
+      @bucket = create(:funded_bucket, user: user, group: group)
+      @bucket_author = @bucket.user
+
+      # create member with $100.00, who has subscribed to participant activity + has commented on bucket
+      @subscribed_participant = create_bucket_participant(bucket: @bucket, subscribed: true)
+
+      # bucket creator comments on their bucket
+      create(:comment, bucket: @bucket, user: @bucket_author)
+
+      # create bucket participant not subscribed to participant activity
```
Comment (mixmix) [inline spec/services/bucket_service_spec.rb]: I think these instances of "not subscribed"   should be replaced with "unsubscribed"   .. i got a little confused to start with. 
not big deal though

File: spec/services/comment_service_spec.rb [hunk 5]
```diff
@@ -22,6 +61,15 @@
       end
     end
 
+    context "bucket author is subcribed to personal activity, but is also commenter" do
```
Comment (mixmix) [inline spec/services/comment_service_spec.rb]: this is already tested up on line#49  - don't bother taking it out though. Even reading this is costing too much!

**Comment by mixmix** [comment]
Reviewed the specs and some of the model changes. 
Your work is really clear, pleasure to review 

:pizza: :pizza: :pizza: :pizza: 

Awesome job

File: app/services/bucket_service.rb (hunk 1 above)
Comment (mixmix) [inline app/services/bucket_service.rb]: you might be quite right. 

mmm

```ruby
scope :published_for_group, ->(group) { joins(:membership).where(memberships: {archived_at: nil, group_id: group.id}) }
```

I think that would work ... it's a fancy join I learnt from Chris T.
Anyway not sure if it's a worthwhile complexity here...

Comment (ahdinosaur) [inline app/services/bucket_service.rb]: :+1: to using model scopes.

File: db/migrate/20151115221849_add_subscribed_to_participant_activity_to_users.rb (hunk 2 above)
Comment (euglazer) [inline db/migrate/20151115221849_add_subscribed_to_participant_activity_to_users.rb]: https://github.com/cobudget/cobudget-api/commit/f09eef71613959fb2ad6f7d3ea6a231989d93e8f

damn took me super long to figure out that you can't `validates_presence_of` on `boolean` properties

File: app/services/bucket_service.rb (hunk 1 above)
Comment (euglazer) [inline app/services/bucket_service.rb]: https://github.com/cobudget/cobudget-api/commit/c13f083d31e63a2f3e7ab309b02b8bb6ff3431ce

**Pull request merged**